rust/agama-server/src/software/web.rs

```diff
@@ -0,0 +1,242 @@
+//! This module implements the web API for the software module.
+//!
+//! The module offers two public functions:
+//!
+//! * `software_service` which returns the Axum service.
+//! * `software_stream` which offers an stream that emits the software events coming from D-Bus.
+
+use crate::{error::Error, web::Event};
+use agama_lib::{
+    error::ServiceError,
+    product::{Product, ProductClient},
+    software::{
+        proxies::{Software1Proxy, SoftwareProductProxy},
+        Pattern, SelectedBy, SoftwareClient,
+    },
+};
+use axum::{
+    extract::State,
+    http::StatusCode,
+    response::{IntoResponse, Response},
+    routing::{get, post, put},
+    Json, Router,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::collections::HashMap;
+use thiserror::Error;
+use tokio_stream::{Stream, StreamExt};
+
+#[derive(Clone)]
+struct SoftwareState<'a> {
+    product: ProductClient<'a>,
+    software: SoftwareClient<'a>,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+struct SoftwareConfig {
+    patterns: Option<Vec<String>>,
+    product: Option<String>,
+}
+
+#[derive(Error, Debug)]
+pub enum SoftwareError {
+    #[error("Service error: {0}")]
+    Error(#[from] ServiceError),
+}
+
+impl IntoResponse for SoftwareError {
+    fn into_response(self) -> Response {
+        let body = json!({
+            "error": self.to_string()
+        });
+        (StatusCode::BAD_REQUEST, Json(body)).into_response()
+    }
+}
+
+/// Returns an stream that emits software related events coming from D-Bus.
+///
+/// * `connection`: D-Bus connection to listen for events.
+pub async fn software_stream(dbus: zbus::Connection) -> Result<impl Stream<Item = Event>, Error> {
+    Ok(StreamExt::merge(
+        product_changed_stream(dbus.clone()).await?,
+        patterns_changed_stream(dbus.clone()).await?,
+    ))
+}
+
+async fn product_changed_stream(
+    dbus: zbus::Connection,
+) -> Result<impl Stream<Item = Event>, Error> {
+    let proxy = SoftwareProductProxy::new(&dbus).await?;
+    let stream = proxy
+        .receive_selected_product_changed()
+        .await
+        .then(|change| async move {
+            if let Ok(id) = change.get().await {
```

Comment:
Can it happen that change result is not `Ok`? And if so, in which case? I would at least document it here.

Reply:
If it fails, it is because it cannot deserialize the information coming from D-Bus. Our proxy could be outdated, for instance. But I do not think we should put a comment about that in each place where we use this function.

See https://docs.rs/zbus/latest/src/zbus/proxy/mod.rs.html#196-200 (because [get documentation](https://docs.rs/zbus/latest/zbus/proxy/struct.PropertyChanged.html#method.get)) is not enough.

In general, any conversion problem is, most probably, because our proxies are outdated.

Reply:
ok, so basically internal problem and not user side.